pinot-broker/src/main/java/org/apache/pinot/broker/broker/helix/ClusterChangeMediator.java

```diff
@@ -18,142 +18,136 @@
+                    TimeUnit.MILLISECONDS);
+                processClusterChange(changeType, changeHandler);
+              } else {
+                long lastProcessTime = _lastProcessTimeMap.get(changeType);
+                if (currentTime - lastProcessTime > PROACTIVE_CHANGE_CHECK_INTERVAL_MS) {
+                  LOGGER.info("Proactive check {} change", changeType);
+                  brokerMetrics.addMeteredGlobalValue(BrokerMeter.PROACTIVE_CLUSTER_CHANGE_CHECK, 1L);
+                  processClusterChange(changeType, changeHandler);
+                }
               }
             }
 
-            if (instanceConfigUpdated) {
-              try {
-                _helixExternalViewBasedRouting.processInstanceConfigChange();
-              } catch (Exception e) {
-                LOGGER.warn("Caught exception while processing instance config", e);
-              }
-            }
+            // Sleep 1 second after each round of checks
+            Thread.sleep(1000L);
           } catch (InterruptedException e) {
-            LOGGER.warn("Was interrupted while waiting for a cluster change", e);
+            LOGGER.warn("Cluster change handling thread is interrupted, stopping the thread");
             break;
+          } catch (Exception e) {
+            LOGGER.error("Caught exception while handling changes", e);
           }
         }
-
-        LOGGER.warn("Stopping deferred cluster state update thread");
-        _deferredClusterUpdater = null;
       }
     };
+    _clusterChangeHandlingThread.start();
+  }
 
-    _deferredClusterUpdater.start();
+  private void processClusterChange(ChangeType changeType, ClusterChangeHandler changeHandler) {
+    long startTime = System.currentTimeMillis();
+    LOGGER.info("Start processing {} change", changeType);
+    changeHandler.processClusterChange();
+    long endTime = System.currentTimeMillis();
+    LOGGER.info("Finish processing {} change in {}ms", changeType, endTime - startTime);
+    _lastProcessTimeMap.put(changeType, endTime);
   }
 
   @Override
   public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
-    // If the deferred update thread is alive, defer the update
-    if (_deferredClusterUpdater != null && _deferredClusterUpdater.isAlive()) {
-      try {
-        _clusterChangeQueue.put(new ImmutablePair<>(UpdateType.EXTERNAL_VIEW, System.currentTimeMillis()));
-      } catch (InterruptedException e) {
-        LOGGER.warn("Was interrupted while trying to add external view change to queue", e);
-      }
-    } else {
-      LOGGER.warn(
-          "Deferred cluster updater thread is null or stopped, not deferring external view routing table rebuild");
-      _helixExternalViewBasedRouting.processExternalViewChange();
-      _tableQueryQuotaManager.processQueryQuotaChange();
-    }
+    Preconditions.checkState(externalViewList.isEmpty(), "Helix pre-fetch should be disabled");
```

Comment:
If I understand right, the pre-fetch is an annotation on this method. Why not add the annotation and set it to be false? Another way of putting this question: What is an admin supposed to do if they see this pre-condition fail?

Reply:
The pre-fetch is an annotation on the class (line 52). The purpose of this preconditions is to prevent the extra ZK accesses if the Helix behavior changes in the future and ignores the annotation. In such case, our test should be able to catch the preconditions check failure.